File: backend/KnxMonitor.Infrastructure/Services/ProjectService.cs
using KnxMonitor.Core.DTOs;
using KnxMonitor.Core.Entities;
using KnxMonitor.Core.Interfaces;
using KnxMonitor.Core.Services;
using Microsoft.EntityFrameworkCore;

namespace KnxMonitor.Infrastructure.Services;

public class ProjectService : IProjectService
{
    private readonly IProjectRepository _projectRepository;
    private readonly IGroupAddressRepository _groupAddressRepository;
    private readonly IKnxProjectParserService _parserService;
    private readonly IGroupAddressCacheService _cacheService;
    private readonly IKnxConnectionService _knxService;
    private readonly IKnxConfigurationRepository _configRepository;

    public ProjectService(
        IProjectRepository projectRepository,
        IGroupAddressRepository groupAddressRepository,
        IKnxProjectParserService parserService,
        IGroupAddressCacheService cacheService,
        IKnxConnectionService knxService,
        IKnxConfigurationRepository configRepository)
    {
        _projectRepository = projectRepository;
        _groupAddressRepository = groupAddressRepository;
        _parserService = parserService;
        _cacheService = cacheService;
        _knxService = knxService;
        _configRepository = configRepository;
    }

    public async Task<ProjectDto> UploadProjectAsync(Stream fileStream, string fileName)
    {
        // Create project entry
        var project = new Project
        {
            Name = Path.GetFileNameWithoutExtension(fileName),
            FileName = fileName,
            ImportDate = DateTime.UtcNow,
            IsActive = false
        };

        await _projectRepository.AddAsync(project);

        try
        {
            // Parse the .knxproj file
            var parsed = await _parserService.ParseProjectFileAsync(fileStream, project.Id);

            // Save parsed data
            project.GroupAddresses = parsed.GroupAddresses;
            project.Devices = parsed.Devices;
            project.Locations = parsed.Locations;
            project.CommunicationObjects = parsed.CommunicationObjects;
            project.GroupRanges = parsed.GroupRanges;
            project.KeyringKeys = parsed.KeyringKeys;

            await _projectRepository.UpdateAsync(project);

            return new ProjectDto
            {
                Id = project.Id,
                Name = project.Name,
                FileName = project.FileName,
                ImportDate = project.ImportDate,
                IsActive = project.IsActive,
                GroupAddressCount = parsed.GroupAddresses.Count,
                DeviceCount = parsed.Devices.Count
            };
        }
        catch (Exception ex)
        {
            // Delete project if parsing failed
            await _projectRepository.DeleteAsync(project);
            throw new InvalidOperationException($"Failed to import project: {ex.Message}", ex);
        }
    }

    public async Task<List<ProjectDto>> GetAllProjectsAsync()
    {
        var projects = await _projectRepository.GetAllWithCountsAsync();

        return projects.Select(p => new ProjectDto
        {
            Id = p.Id,
            Name = p.Name,
            FileName = p.FileName,
            ImportDate = p.ImportDate,
            IsActive = p.IsActive,
            GroupAddressCount = p.GroupAddresses.Count,
            DeviceCount = p.Devices.Count
        }).ToList();
    }

    public async Task<ProjectDetailsDto?> GetProjectDetailsAsync(int id)
    {
        var project = await _projectRepository.GetWithDetailsAsync(id);

        if (project == null)
            return null;

        return new ProjectDetailsDto
        {
            Id = project.Id,
            Name = project.Name,
            FileName = project.FileName,
            ImportDate = project.ImportDate,
            IsActive = project.IsActive,
            GroupAddresses = project.GroupAddresses.Select(ga => new GroupAddressDto
            {
                Id = ga.Id,
                Address = ga.Address,
                Name = ga.Name,
                Description = ga.Description,
                DatapointType = ga.DatapointType
            }).ToList(),
            Devices = project.Devices.Select(d => new DeviceDto
            {
                Id = d.Id,
                Name = d.Name,
                PhysicalAddress = d.PhysicalAddress,
                Manufacturer = d.Manufacturer,
                ProductName = d.ProductName
            }).ToList()
        };
    }

    public async Task<bool> ActivateProjectAsync(int id)
    {
        var project = await _projectRepository.GetByIdAsync(id);
        if (project == null)
            return false;

        // Atomically make exactly this project active (set-based, single transaction).
        await _projectRepository.SetActiveExclusiveAsync(id);

        // Refresh the cache with the new active project
        await _cacheService.RefreshAsync();

        // Couple the bus to the active project (model: active ⇒ connected). If auto-connect is
        // enabled, (re)connect now — this also clears any manual-disconnect latch left by a
        // previous deactivate, so re-activating restores the live link.
        var configs = await _configRepository.GetAllAsync();
        var config = configs.FirstOrDefault(c => c.IsActive) ?? configs.FirstOrDefault();
        if (config != null && config.AutoConnect && !_knxService.IsConnected)
        {
            try { await _knxService.ConnectAsync(config); }
            catch { /* non-fatal: the auto-connect worker will retry */ }
        }

        return true;
    }

    public async Task<bool> DeactivateProjectAsync(int id)
    {
        var project = await _projectRepository.GetByIdAsync(id);
        if (project == null)
            return false;

        if (!project.IsActive)
            return true; // already inactive — nothing to do

        project.IsActive = false;
        await _projectRepository.UpdateAsync(project);

        // No active project anymore: clear the GA cache and sever the bus link. DisconnectAsync
        // latches ManualDisconnect, so the auto-connect worker won't immediately reconnect.
        await _cacheService.RefreshAsync();
        await _knxService.DisconnectAsync();

        return true;
    }

    public async Task<bool> DeleteProjectAsync(int id)
    {
        var project = await _projectRepository.GetByIdAsync(id);
        if (project == null)
            return false;

        var wasActive = project.IsActive;

        // Set-based bulk delete (telegrams preserved, GA refs nulled in one indexed UPDATE).
        var deleted = await _projectRepository.DeleteProjectFastAsync(id);
        if (!deleted)
            return false;

        // If the deleted project was active, no project is active anymore: clear cache + drop bus.
        if (wasActive)
        {
            await _cacheService.RefreshAsync();
            await _knxService.DisconnectAsync();
        }

        return true;
    }

    public async Task<ProjectDeletePreviewDto?> GetDeletePreviewAsync(int id)
    {
        var project = await _projectRepository.GetWithDetailsAsync(id);
        if (project == null)
            return null;

        return new ProjectDeletePreviewDto
        {
            GroupAddressCount = project.GroupAddresses.Count,
            DeviceCount = project.Devices.Count,
            TelegramCount = await _projectRepository.CountTelegramsForProjectAsync(id)
        };
    }

    public async Task<List<LocationDto>?> GetLocationsAsync(int id)
    {
        if (!await _projectRepository.ExistsAsync(id))
            return null;

        var locations = await _projectRepository.GetLocationsAsync(id);

        return locations.Select(l => new LocationDto
        {
            Id = l.Id,
            ExternalId = l.ExternalId,
            Name = l.Name,
            Type = l.Type,
            ParentExternalId = l.ParentExternalId,
            DeviceAddresses = SplitList(l.DeviceAddresses),
            GroupAddresses = SplitList(l.GroupAddresses)
        }).ToList();
    }

    public async Task<List<CommunicationObjectDto>?> GetCommObjectsAsync(int id, string? groupAddress)
    {
        if (!await _projectRepository.ExistsAsync(id))
            return null;

        var comObjects = await _projectRepository.GetCommObjectsAsync(id, groupAddress);

        // One lookup dict (DeviceAddress -> device) for the whole call, so we can resolve the device
        // name/manufacturer/product behind each com object's physical address without an extra query per row.
        var devicesByAddress = (await _projectRepository.GetDevicesAsync(id))
            .Where(d => !string.IsNullOrEmpty(d.PhysicalAddress))
            .GroupBy(d => d.PhysicalAddress)
            .ToDictionary(g => g.Key, g => g.First());

        return comObjects.Select(c =>
        {
            devicesByAddress.TryGetValue(c.DeviceAddress, out var device);
            return new CommunicationObjectDto
            {
                Id = c.Id,
                DeviceAddress = c.DeviceAddress,
                DeviceName = device?.Name,
                DeviceManufacturer = device?.Manufacturer,
                DeviceProductName = device?.ProductName,
                Number = c.Number,
                Name = c.Name,
                FunctionText = c.FunctionText,
                GroupAddressLink = c.GroupAddressLink,
                DatapointType = c.DatapointType,
                Flags = c.Flags
            };
        }).ToList();
    }

    public async Task<List<GroupRangeDto>?> GetGroupRangesAsync(int id)
    {
        if (!await _projectRepository.ExistsAsync(id))
            return null;

        var ranges = await _projectRepository.GetGroupRangesAsync(id);

        return ranges.Select(r => new GroupRangeDto
        {
            Name = r.Name,
            RangeStart = r.RangeStart,
            RangeEnd = r.RangeEnd
        }).ToList();
    }

    public async Task<DeviceDto?> GetDeviceByAddressAsync(int projectId, string physicalAddress)
    {
        if (!await _projectRepository.ExistsAsync(projectId))
            return null;

        var device = await _projectRepository.GetDeviceByAddressAsync(projectId, physicalAddress);
        if (device == null)
            return null;

        return new DeviceDto
        {
            Id = device.Id,
            Name = device.Name,
            PhysicalAddress = device.PhysicalAddress,
            Manufacturer = device.Manufacturer,
            ProductName = device.ProductName
        };
    }

    public async Task<KeyringUploadResultDto?> UploadKeyringAsync(int id, byte[] keyringData, string keyringPassword)
    {
        if (!await _projectRepository.ExistsAsync(id))
            return null;

        // Decrypt + map (throws InvalidOperationException on bad password/format — surfaced by controller).
        var keys = await _parserService.ParseKeyringAsync(keyringData, keyringPassword, id);

        // Replace any previously stored keyring material for this project.
        await _projectRepository.ReplaceKeyringKeysAsync(id, keys);

        // Also persist the RAW keyring bytes + password so KNX Data Secure can Load() them at
        // connect time (the decrypted per-GA keys above are not sufficient for that Load() call).
        // SECURITY: stores the keyring password at rest — see ProjectKeyringBlob.
        await _projectRepository.UpsertKeyringBlobAsync(id, keyringData, keyringPassword);

        return new KeyringUploadResultDto
        {
            TotalKeys = keys.Count,
            GroupAddressKeys = keys.Count(k => k.KeyType == "GroupAddress"),
            ToolKeys = keys.Count(k => k.KeyType == "ToolKey"),
            HasBackboneKey = keys.Any(k => k.KeyType == "Backbone")
        };
    }

    private static List<string> SplitList(string? csv) =>
        string.IsNullOrWhiteSpace(csv)
            ? new List<string>()
            : csv.Split(',', StringSplitOptions.RemoveEmptyEntries).ToList();
}

File: frontend/src/app/core/services/project.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment.development';
import { ImportJob, ProvideInput } from '../../shared/models/import-job.model';

export interface ProjectDto {
  id: number;
  name: string;
  fileName: string;
  importDate: Date;
  isActive: boolean;
  groupAddressCount: number;
  deviceCount: number;
}

export interface GroupAddressDto {
  id: number;
  address: string;
  name: string;
  description?: string;
  datapointType?: string;
}

export interface DeviceDto {
  id: number;
  name: string;
  physicalAddress: string;
  manufacturer?: string;
  productName?: string;
}

export interface ProjectDetailsDto {
  id: number;
  name: string;
  fileName: string;
  importDate: Date;
  isActive: boolean;
  groupAddresses: GroupAddressDto[];
  devices: DeviceDto[];
}

export interface ProjectDeletePreview {
  groupAddressCount: number;
  deviceCount: number;
  /** Recorded telegrams mapped to this project; kept on delete but lose their name mapping. */
  telegramCount: number;
}


/** Flat node of the building/location tree (rebuilt client-side via externalId/parentExternalId). */
export interface LocationDto {
  id: number;
  externalId: string;
  name: string;
  type: string;
  parentExternalId?: string | null;
  deviceAddresses: string[];
  groupAddresses: string[];
}

export interface CommunicationObjectDto {
  id: number;
  deviceAddress: string;
  deviceName?: string | null;
  number: number;
  name?: string | null;
  functionText?: string | null;
  groupAddressLink: string;
  datapointType?: string | null;
  flags?: string | null;
}

/** A KNX GroupRange = main/middle group name spanning a contiguous block of GA addresses. */
export interface GroupRangeDto {
  name: string;
  rangeStart: number;
  rangeEnd: number;
}

export interface KeyringUploadResultDto {
  totalKeys: number;
  groupAddressKeys: number;
  toolKeys: number;
  hasBackboneKey: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class ProjectService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/projects`;

  uploadProject(file: File): Observable<ImportJob> {
    const formData = new FormData();
    formData.append('file', file, file.name);
    return this.http.post<ImportJob>(`${this.apiUrl}/upload`, formData);
  }

  getImportStatus(jobId: string): Observable<ImportJob> {
    return this.http.get<ImportJob>(`${this.apiUrl}/imports/${jobId}`);
  }

  provideInput(jobId: string, input: ProvideInput): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(`${this.apiUrl}/imports/${jobId}/provide-input`, input);
  }

  cancelImport(jobId: string): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/imports/${jobId}`);
  }

  getAllProjects(): Observable<ProjectDto[]> {
    return this.http.get<ProjectDto[]>(this.apiUrl);
  }

  getProjectDetails(id: number): Observable<ProjectDetailsDto> {
    return this.http.get<ProjectDetailsDto>(`${this.apiUrl}/${id}`);
  }

  activateProject(id: number): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(`${this.apiUrl}/${id}/activate`, {});
  }

  deactivateProject(id: number): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(`${this.apiUrl}/${id}/deactivate`, {});
  }

  getDeletePreview(id: number): Observable<ProjectDeletePreview> {
    return this.http.get<ProjectDeletePreview>(`${this.apiUrl}/${id}/delete-preview`);
  }

  deleteProject(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/${id}`);
  }

  // --- Locations (building tree) / communication objects / keyring -----------

  getLocations(projectId: number): Observable<LocationDto[]> {
    return this.http.get<LocationDto[]>(`${this.apiUrl}/${projectId}/locations`);
  }

  getGroupRanges(projectId: number): Observable<GroupRangeDto[]> {
    return this.http.get<GroupRangeDto[]>(`${this.apiUrl}/${projectId}/groupranges`);
  }

  getCommObjects(projectId: number, address?: string): Observable<CommunicationObjectDto[]> {
    const query = address ? `?address=${encodeURIComponent(address)}` : '';
    return this.http.get<CommunicationObjectDto[]>(`${this.apiUrl}/${projectId}/commobjects${query}`);
  }

  /** Resolves a single device of a project by its exact physical address (e.g. "1.0.59"). */
  getDeviceByAddress(projectId: number, address: string): Observable<DeviceDto> {
    return this.http.get<DeviceDto>(`${this.apiUrl}/${projectId}/device?address=${encodeURIComponent(address)}`);
  }

  uploadKeyring(projectId: number, file: File, password: string): Observable<KeyringUploadResultDto> {
    const formData = new FormData();
    formData.append('file', file, file.name);
    formData.append('password', password);
    return this.http.post<KeyringUploadResultDto>(`${this.apiUrl}/${projectId}/keyring`, formData);
  }
}
